cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -145,20 +153,54 @@ public String call(DatasetContext context) throws Exception {
      */
     long processNotifications() {
       try {
-        final MessageFetcher fetcher = messagingContext.getMessageFetcher();
-        String lastFetchedMessageId = Transactionals.execute(transactional, new TxCallable<String>() {
-          @Override
-          public String call(DatasetContext context) throws Exception {
-            return fetchAndProcessNotifications(context, fetcher);
+        String lastFetchedMessageId = null;
+        // Gather a local copy of fetched notifications to batch process in one transaction
+        List<Notification> fetchedNotifications = new ArrayList<>();
+        try (CloseableIterator<Message> iterator = messagingContext.getMessageFetcher()
+                                                                   .fetch(NamespaceId.SYSTEM.getNamespace(), topic,
+                                                                          100, messageId)) {
+          LOG.trace("Fetch with messageId = {}", messageId);
+          while (iterator.hasNext() && !stopping) {
+            Message message = iterator.next();
+            lastFetchedMessageId = message.getId();
```

Comment:
Actually, I think its easier to update it here. This also needs to get updated when the notification can't be parsed (to skip it in the next iteration), so if we didn't update it in the loop then, the last element in the list represents the last parsed notification, not the last message overall (which could have been invalid).